power_grid_model_c/power_grid_model/include/power_grid_model/topology.hpp

```diff
@@ -495,41 +495,61 @@ class Topology {
 
     // proxy class to find the coupled object in math model in the coupling process to a single type object
     //    given a particular component index
-    struct SingleTypeObjectFinder {
-        Idx size() const { return static_cast<Idx>(component_obj_idx.size()); }
-        Idx2D find_math_object(Idx component_i) const { return objects_coupling[component_obj_idx[component_i]]; }
-        IdxVector const& component_obj_idx;
-        std::vector<Idx2D> const& objects_coupling;
+    class SingleTypeObjectFinder {
+      public:
+        SingleTypeObjectFinder(IdxVector const& component_obj_idx, std::vector<Idx2D> const& objects_coupling)
```

Comment:
Why didn't we have a constructor in the first place? Was it constructed somewhere else?

Reply:
initially, this was a `struct` that would be constructed with a `IdxVector const&` to `component_obj_idx`.

The problem is, that this breaks `std::movable` behaviour (`const` and `&` member fields cause `move` constructor to be implicitly deleted, because `const` and l-value `&` objects can't be moved, only copied)

Just replacing `const&` with `const*` would fix this, BUT it would imply that a `nullptr` check would have to be done at every function execution (remember that `struct` by convention implies that the objects of that type do not have so-called invariants, meaning that not all combinations of member variables are valid, while for `class`, some values/combinations of member variables may be invalid by design).

In this case, we do not want to check for `nullptr` every executable, so I:
* made this `struct` a `class`
* added a constructor that takes in a `const&` that ensures that `nullptr` will never be set as a value
* made the members `private` instead of `public`
This ensures that the variable will never ever have a `nullptr` and therefore, the `nullptr` check is not necessary

Reply:
Thanks! I missed the fact that it was a struct before.